ENH: Add Inverse of Log CDF of Normal Distribution

steppi opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Closes gh-13923. One step towards closing gh-11465.
#### What does this implement/fix?
<!--Please explain your changes.-->
This PR adds an implementation of the inverse of the Log CDF of the Normal Distribution which offers more precision than applying `ndtri(exp(y))` when `y` is very small or close to zero. For instance, `ndtri(exp(y))` saturates at `-inf` when `exp(y)` underflows for `y < ~ -745` and saturates at `inf` when `exp(y) == 1.0` for `-1e-16 - epsilon < y < 0`. The function `ndtri_exp` added in this PR returns a non-infinite answer for all floating point numbers y with `-inf < y < 0`. 
<!--Any additional information you think is important.-->
[The underlying C code for `scipy.special.ndtri`](https://github.com/scipy/scipy/blob/master/scipy/special/cephes/ndtri.c) already contains an excellent approximation to the inverse of the log CDF of the normal distribution which is used to compute `ndtri(x)` for `0 < x < exp(-2)` (and `1 - exp(-2) < x < 1`) by symmetry, by taking the logarithm of x and then applying the approximation. This implementation is just not exposed as a way to compute the inverse log CDF. The implementation proposed in this PR makes use of this approximation directly for `y < -2`. 

The name `ndtri_exp` was a tossed off suggestion by @mdhaber and I've become convinced it is better than all of the other obvious possibilities, but I'd be happy to hear other suggestions. It seems natural to add a corresponding `log_ppf` method to `scipy.stats.norm`, but this was not done in this PR to avoid introducing an API inconsistency where some stats distribution classes have `log_ppf` methods and others don't.

person142 commented:
Hey @steppi-I’m going to have very little time to review in the next two weeks, but I’ll get to this after that. Ping me if I forget.

steppi commented:
Thanks @person142, will do. Should I worry about some of the windows builds timing out?

person142 commented:
Probably not; won’t stop review, and if they’re timing out after that we’ll figure it out.

tylerjereddy commented on scipy/special/add_newdocs.py:
```diff
@@ -9643,3 +9643,39 @@ def add_newdoc(name, doc):
+
+    Examples
+    --------
+    >>> import numpy as np
```
I think this line is implicit in our docstring examples

tylerjereddy commented on scipy/special/_ndtri_exp.pxd:
different -> difference, I think

lispsil commented on scipy/special/_ndtri_exp.pxd:
`log1p(-exp(-2))` maybe this should be a named constant to?

steppi replied:
That's a good idea. I'm not sure what to call it yet but can change it when I think of a good name.

zoj613 replied:
probably doesn't need a name. Could just plug in the value and add an inline comment with the full expression.
```cython
elif y > 0.345423545343434:  # log1p(-exp(-2))
   ...
```

steppi commented:
I found and fixed a bug for an edge case where `-DBL_MAX < y < -1/2 * DBL_MAX`.  The function was returning `nan` instead of `-inf` for this case

steppi commented:
Oh, it looks like I should rewrite the tests to use [assert_func_equal](https://github.com/scipy/scipy/blob/master/scipy/special/_testutils.py#L53) from `_testutils.py`. I'll have time to do that over the weekend.

person142 commented on scipy/special/tests/test_ndtri_exp.py:
```diff
@@ -0,0 +1,61 @@
+    def test_very_small_arg(self, test_input, uniform_random_points):
+        scale = test_input
+        points = scale * (0.5 * uniform_random_points + 0.5)
+        assert_func_equal(log_ndtr_ndtri_exp, lambda y: y, points,
```
You'll want to flip on `nan_ok=True` on all these calls.

person142 reviewed (commented):
Looks good overall; some small comments.

steppi commented on scipy/special/_ndtri_exp.pxd:
Thank you! I wasn't sure how to compactly initialize the arrays while the gil is released.

steppi commented:
@person142 Thanks for the thorough review. You've taught me some useful stuff. I've think I've addressed all of your comments in the commits I just pushed.

steppi commented on scipy/special/tests/test_ndtri_exp.py:
```diff
@@ -0,0 +1,61 @@
+
+@pytest.fixture(scope='class')
+def uniform_random_points():
+    random_state = np.random.RandomState(1729)
```
Ah, OK, that's reasonable. I'll keep the math culture in-jokes to my own projects.

steppi commented:
@tylerjereddy @person142  What's the next step? Is it possible to get this into 1.7.0? It's reasonable if not since time is tight.

zoj613 commented on scipy/special/_ndtri_exp.pxd:
```diff
@@ -0,0 +1,174 @@
+For 0 < p < exp(-2), the cephes implementation of ndtri uses an approximation
+for ndtri(p) which is a function of z = sqrt(-2.0 * log(p)). Letting
+y = log(p), for y < -2, ndtri_exp uses this approximation in log(p) directly.
+This allows this implementation to achieve high precision for very small y,
+whereas ndtri(exp(y)) becomes inf when exp(y) underflows for y < ~ -745.1.
```
```suggestion
This allows the implementation to achieve high precision for very small y,
whereas ndtri(exp(y)) evaluates to infinity. This is because exp(y) underflows for y < ~ -745.1.
```
I think being explicit with the use of "infinity" is clearer than just writing it as "inf" in the docstring.

zoj613 commented on scipy/special/_ndtri_exp.pxd:
I think breaking up this sentence into two will help make easier to read.

zoj613 commented on scipy/special/_ndtri_exp.pxd:
```suggestion
The case

	-2 <= y <= log(1 - exp(-2)) ~ -0.1454

corresponds to
	~ 0.135 <= p <= ~ 0.865.
```

zoj613 commented on scipy/special/_ndtri_exp.pxd:
```suggestion
The derivative of ndtri is sqrt(2 * pi) * exp(ndtri(x)**2 / 2).
It is ~4.597 at x ~ 0.135, decreases monotonically to \sqrt(2 * pi) ~ 2.507 at x = 0,
and increases monotonically again to ~4.597 at x ~ 0.865.
```

zoj613 commented on scipy/special/_ndtri_exp.pxd:
```suggestion
It can be checked that all higher derivatives follow a similar pattern.
Their absolute value takes on a maximum (for this interval) at x ~ 0.135,
```

zoj613 commented on scipy/special/_ndtri_exp.pxd:
```suggestion
Assuming that exp is accurate within +-ULP (unit of least precision),
we observe a value of at most ~6.0474e-16 for

	abs(ndtri(x + epsilon) - ndtri(x))

if x is near exp(-2) or 1 - exp(-2) and epsilon is equal to the unit of least
precision of x.
```

zoj613 commented on scipy/special/tests/test_ndtri_exp.py:
```diff
@@ -0,0 +1,75 @@
+
+
+class TestNdtriExp:
+    """Tests that ndtri_exp is sufficiently close to an inverse of log_ndtr
```
```suggestion
    """Tests that ndtri_exp is sufficiently close to an inverse of log_ndtr.
```

zoj613 commented on scipy/special/tests/test_ndtri_exp.py:
```diff
@@ -0,0 +1,75 @@
+class TestNdtriExp:
+    """Tests that ndtri_exp is sufficiently close to an inverse of log_ndtr
+
+    Separate tests for the five intervals (-inf, -10),
+    [-10, -2), [-2, -0.14542), [-0.14542, -1e-6), [-1e-6, 0).
+    ndtri_exp(y) is computed in three different ways depending on if y
+    is in (-inf, -2), [-2, log(1 - exp(-2))], [log(1 - exp(-2), 0).
+    Each of these intervals is given its own test with two additional tests
+    for handling very small values and values very close to zero.
```
```suggestion
    We have separate tests for the five intervals (-inf, -10),
    [-10, -2), [-2, -0.14542), [-0.14542, -1e-6) and [-1e-6, 0).
    ndtri_exp(y) is computed in three different ways depending on if y
    is in (-inf, -2), [-2, log(1 - exp(-2))] or [log(1 - exp(-2), 0).
    Each of these intervals is given its own test with two additional tests
    for handling very small values and values very close to zero.
```

steppi commented:
Thanks for the proofreading @zoj613. Sorry, I should have paid more attention to grammar and formatting.

person142 commented:
Re making it into 1.7-I can probably (?) look at it again tonight. Can’t remember the exact 1.7 timeline.

person142 commented on scipy/special/_ndtri_exp.pxd:
Could remove the comment now; when using `M_SQRT2` it's clear what it does.

person142 reviewed (approved):
Looks good-CI still running; I will try to remember to pop back in and merge later tonight.

Ping me if I forget.

steppi commented:
Thanks @person142.

person142 commented:
Thanks @steppi!
